src/background.c

```diff
@@ -0,0 +1,348 @@
+                                                   &error))
+    {
+      g_dbus_error_strip_remote_error (error);
+      g_debug ("No background permissions found: %s", error->message);
+      return UNSET;
+    }
+
+  if (!g_variant_lookup (out_perms, app_id, "^a&s", &permissions))
+    {
+      g_debug ("No permissions stored for: app %s", app_id);
+
+      return UNSET;
+    }
+  else if (g_strv_length ((char **)permissions) != 1)
+    {
+      g_autofree char *a = g_strjoinv (" ", (char **)permissions);
+      g_warning ("Wrong permission format, ignoring (%s)", a);
+      return UNSET;
+    }
+  g_debug ("permission store: app %s -> %s", app_id, permissions[0]);
+
+  if (strcmp (permissions[0], "yes") == 0)
+    return YES;
+  else if (strcmp (permissions[0], "no") == 0)
+    return NO;
+  else if (strcmp (permissions[0], "ask") == 0)
+    return ASK;
+  else
+    {
+      g_autofree char *a = g_strjoinv (" ", (char **)permissions);
+      g_warning ("Wrong permission format, ignoring (%s)", a);
+    }
+
+  return UNSET;
+}
+
+static void
+set_permission (const char *app_id,
+                Permission permission)
+{
+  g_autoptr(GError) error = NULL;
+  const char *permissions[2];
+
+  if (permission == ASK)
+    permissions[0] = "ask";
+  else if (permission == YES)
+    permissions[0] = "yes";
+  else if (permission == NO)
+    permissions[0] = "no";
+  else
+    {
+      g_warning ("Wrong permission format, ignoring");
+      return;
+    }
+  permissions[1] = NULL;
+
+  if (!xdp_impl_permission_store_call_set_permission_sync (get_permission_store (),
+                                                           PERMISSION_TABLE,
+                                                           TRUE,
+                                                           PERMISSION_ID,
+                                                           app_id,
+                                                           (const char * const*)permissions,
+                                                           NULL,
+                                                           &error))
+    {
+      g_dbus_error_strip_remote_error (error);
+      g_warning ("Error updating permission store: %s", error->message);
+    }
+}
+
+static void
+handle_request_background_in_thread_func (GTask *task,
+                                          gpointer source_object,
+                                          gpointer task_data,
+                                          GCancellable *cancellable)
+{
+  Request *request = (Request *)task_data;
+  GVariant *options;
+  const char *app_id;
+  Permission permission;
+  gboolean allowed;
+  const char *reason = NULL;
+
+  REQUEST_AUTOLOCK (request);
+
+  options = (GVariant *)g_object_get_data (G_OBJECT (request), "options");
+  g_variant_lookup (options, "reason", "s", &reason);
+
+  app_id = xdp_app_info_get_id (request->app_info);
+  permission = get_permission (app_id);
```

Comment:
I think this needs to special case the xdp_app_info_is_host() case and always return TRUE, because we want to support using portals unconditionally while "just working" when not sandboxed.

Reply:
Probably pass the app_info instead of the app_id into get_permission and have the special case there.